sys/riscv/trap.c

```diff
@@ -0,0 +1,301 @@
+    goto fault;
+  }
+
+  if (!vm_page_fault(vmap, vaddr, access))
+    return;
+
+fault:
+  if (td->td_onfault) {
+    /* Handle copyin/copyout faults. */
+    _REG(ctx, PC) = td->td_onfault;
+  } else if (user_mode_p(ctx)) {
+    /* Send a segmentation fault signal to the user program. */
+    sig_trap(ctx, SIGSEGV);
+  } else {
+    /* Panic when kernel-mode thread uses wrong pointer. */
+    kernel_oops(ctx);
+  }
+}
+
+/*
+ * RISC-V syscall ABI:
+ *  - a7: code
+ *  - a0-5: args
+ *
+ * NOTE: the following code assumes all arguments to syscalls are passed
+ * via registers.
+ */
+static_assert(SYS_MAXSYSARGS <= FUNC_MAXREGARGS - 1,
+              "Syscall args don't fit in registers!");
+
+static void syscall_handler(mcontext_t *uctx, syscall_result_t *result) {
+  /* TODO: eventually we should have a platform-independent syscall handler. */
+  register_t args[SYS_MAXSYSARGS];
+  register_t code = _REG(uctx, A7);
+
+  memcpy(args, &_REG(uctx, A0), sizeof(args));
+
+  if (code > SYS_MAXSYSCALL) {
+    args[0] = code;
+    code = 0;
+  }
+
+  sysent_t *se = &sysent[code];
+  size_t nargs = se->nargs;
+
+  assert(nargs <= SYS_MAXSYSCALL);
+
+  thread_t *td = thread_self();
+  register_t retval = 0;
+
+  assert(td->td_proc);
+
+  int error = se->call(td->td_proc, (void *)args, &retval);
+
+  result->retval = error ? -1 : retval;
+  result->error = error;
+}
+
+#if FPU
+static bool fpu_handler(mcontext_t *uctx) {
+  thread_t *td = thread_self();
+
+  if (td->td_pflags & TDP_FPUINUSE)
+    return false;
+
+  /*
+   * May be a FPE trap. Enable FPE usage
+   * for this thread and try again.
+   */
+  bzero(uctx->__fregs, sizeof(__fregset_t));
+  _REG(uctx, SR) &= ~SSTATUS_FS_MASK;
+  _REG(uctx, SR) |= SSTATUS_FS_CLEAN;
+  td->td_pflags |= TDP_FPUINUSE;
+
+  return true;
+}
+#endif
+
+static void user_trap_handler(ctx_t *ctx) {
+  /*
+   * We came here from user-space, hence interrupts and preemption must
+   * have been enabled.
+   */
+  cpu_intr_enable();
+
+  assert(!intr_disabled() && !preempt_disabled());
+
+  syscall_result_t result;
+  unsigned code = ctx_code(ctx);
+
+  switch (code) {
+    case SCAUSE_INST_PAGE_FAULT:
+    case SCAUSE_LOAD_PAGE_FAULT:
+    case SCAUSE_STORE_PAGE_FAULT:
+      page_fault_handler(ctx);
+      break;
+
+      /* Access fault */
+    case SCAUSE_INST_ACCESS_FAULT:
+    case SCAUSE_LOAD_ACCESS_FAULT:
+    case SCAUSE_STORE_ACCESS_FAULT:
+      /* Missaligned access */
+    case SCAUSE_INST_MISALIGNED:
+    case SCAUSE_LOAD_MISALIGNED:
+    case SCAUSE_STORE_MISALIGNED:
+      sig_trap(ctx, SIGBUS);
+      break;
+
+    case SCAUSE_ECALL_USER:
+      syscall_handler((mcontext_t *)ctx, &result);
+      break;
+
+    case SCAUSE_ILLEGAL_INSTRUCTION: {
+#if FPU
+      if (fpu_handler((mcontext_t *)ctx))
+        break;
+#endif
+      void *epc = (void *)_REG(ctx, PC);
+      klog("%s at %p!", exceptions[code], epc);
+      sig_trap(ctx, SIGILL);
+      break;
+    }
```

Comment:
It can be not clear why `fpu_handler` is a solution for `ILLEGAL_INSTRUCTION`. Please add comment for future readers ;)

Reply:
Done. The comment is placed before the definition of `fpu_handler`.